toolchain/check/testdata/interop/cpp/impls/destroy.carbon

```diff
@@ -0,0 +1,256 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/interop/cpp/impls/destroy.carbon
+
+// --- types.h
+
+class PublicDestructor {
+ public:
+  ~PublicDestructor();
+};
+
+class TrivialDestructor {
+ public:
+  ~TrivialDestructor() = default;
+};
+
+class DeletedDestructor {
+ public:
+  ~DeletedDestructor() = delete;
+};
+
+template<int N = 0>
+class AmbiguousDestructorT {
+ public:
+  ~AmbiguousDestructorT() requires (N != 1);
+  ~AmbiguousDestructorT() requires (N != 2);
+};
+
+using AmbiguousDestructor = AmbiguousDestructorT<>;
+
+class PrivateDestructor {
+ private:
+  ~PrivateDestructor();
+};
+
+class ProtectedDestructor {
+ protected:
+  ~ProtectedDestructor();
+};
+
+// --- destroy_destroyable.carbon
+
+library "[[@TEST_NAME]]";
+
+import Cpp library "types.h";
+
+fn PublicDestroy() {
+  //@dump-sem-ir-begin
+  var a: Cpp.PublicDestructor = {};
+  //@dump-sem-ir-end
+}
+
+fn TrivialDestroy() {
+  //@dump-sem-ir-begin
+  var a: Cpp.TrivialDestructor = {};
+  //@dump-sem-ir-end
+}
+
+// --- fail_destroy_nondestroyable.carbon
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE+16]]:10: in file included here [InCppInclude]
+// CHECK:STDERR: ./types.h:18:7: error: destructor of class 'AmbiguousDestructorT<>' is ambiguous [CppInteropParseError]
+// CHECK:STDERR:    18 | class AmbiguousDestructorT {
+// CHECK:STDERR:       |       ^
+// CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE+12]]:10: in file included here [InCppInclude]
+// CHECK:STDERR: ./types.h:18:7: note: in instantiation of template class 'AmbiguousDestructorT<>' requested here [CppInteropParseNote]
+// CHECK:STDERR:    18 | class AmbiguousDestructorT {
+// CHECK:STDERR:       |       ^
+// CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE+8]]:10: in file included here [InCppInclude]
+// CHECK:STDERR: ./types.h:20:3: note: candidate function [CppInteropParseNote]
+// CHECK:STDERR:    20 |   ~AmbiguousDestructorT() requires (N != 1);
+// CHECK:STDERR:       |   ^
+// CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE+4]]:10: in file included here [InCppInclude]
+// CHECK:STDERR: ./types.h:21:3: note: candidate function [CppInteropParseNote]
+// CHECK:STDERR:    21 |   ~AmbiguousDestructorT() requires (N != 2);
+// CHECK:STDERR:       |   ^
+import Cpp library "types.h";
+
+fn DeletedDestroy() {
+  var a: Cpp.DeletedDestructor = {};
+}
+
+fn AmbiguousDestroy() {
+  // CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE+12]]:10: note: while completing C++ type `Cpp.AmbiguousDestructorT` [InCppTypeCompletion]
+  // CHECK:STDERR:   var a: Cpp.AmbiguousDestructor = {};
+  // CHECK:STDERR:          ^~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_destroy_nondestroyable.carbon:[[@LINE-8]]:3: error: attempt to use a deleted function [CppInteropParseError]
```

Comment:
The sorting here is confusing. Maybe just split up the deleted vs ambiguous test cases into separate splits?